File: java/meetup/src/main/java/Meetup.java
import java.time.DayOfWeek;
import java.time.LocalDate;
import java.time.temporal.TemporalAdjusters;
import java.util.Collections;
import java.util.List;
import java.util.function.Function;

public class Meetup {

    private final LocalDate firstDayOfMonth;

    public Meetup(int month, int year) {
        firstDayOfMonth = LocalDate.of(year, month, 1);
    }

    public LocalDate day(DayOfWeek aimedDayOfWeek, MeetupSchedule meetupSchedule) {
        switch (meetupSchedule) {
            case LAST:
                return firstDayOfMonth.with(TemporalAdjusters.lastInMonth(aimedDayOfWeek));
            case TEENTH:
                LocalDate current = firstDayOfMonth;
                while (!isTenth(current)) {
                    current = next(current, aimedDayOfWeek);
                }
                return current;
            default:

                final List<Function<LocalDate, LocalDate>> necessaryDateTransformations = Collections.nCopies(
                        computeNumberOfShift(meetupSchedule, aimedDayOfWeek),
                        localDate -> next(localDate, aimedDayOfWeek)
                );
                return necessaryDateTransformations
                        .stream()
                        .reduce(firstDayOfMonth, (d, f) -> f.apply(d), (previous, next) -> next);
        }
    }

    private int computeNumberOfShift(MeetupSchedule meetupSchedule, DayOfWeek aimedDayOfWeek) {
        int shift = meetupSchedule.ordinal();
        if (firstDayOfMonth.getDayOfWeek() != aimedDayOfWeek) {
            return shift+1;
        }
        return shift;
    }

    private LocalDate next(LocalDate current, DayOfWeek dayOfWeek) {
        return current.with(TemporalAdjusters.next(dayOfWeek));
    }

    private boolean isTenth(LocalDate localDate) {
        return localDate.getDayOfMonth() > 12 && localDate.getDayOfMonth() < 20;
    }

}

File: clojure/meetup/src/meetup.clj
(ns meetup
  (:import
    (java.time DayOfWeek YearMonth)))

(def day-table {:monday DayOfWeek/MONDAY
                :tuesday DayOfWeek/TUESDAY
                :wednesday DayOfWeek/WEDNESDAY
                :thursday DayOfWeek/THURSDAY
                :friday DayOfWeek/FRIDAY
                :saturday DayOfWeek/SATURDAY
                :sunday DayOfWeek/SUNDAY})

(def teens (set (range 13 20)))

(defn get-day-by-ordinal [days ordinal]
  (case ordinal
    :first (first days)
    :second (second days)
    :third (nth days 2)
    :fourth (nth days 3)
    :last (last days)
    :teenth (some teens days)))

(defn meetup [month year day ordinal]
  (let [year-month (YearMonth/of year month)
        target-day (day-table day)
        month-length (.lengthOfMonth year-month)
        days-in-month
        (->> (range 1 (inc month-length))
             (map (fn [d] (. year-month (atDay d))))
             (filter (fn [d] (= target-day (.getDayOfWeek d))))
             (map #(.getDayOfMonth %)))]
    (conj [year month] (get-day-by-ordinal days-in-month ordinal))))
